### PR TITLE
Broadcast filters

### PR DESCRIPTION
**Pull request by terpstra**
This adds the ability to add Probe filters to the Broadcast coherence manager.

**Type of change**: feature request
**Impact**: API addition (no impact on existing code)
**Development Phase**: implementation

### CODE REVIEW

File: src/main/scala/tilelink/Broadcast.scala [hunk 1]
```diff
@@ -159,119 +184,214 @@ class TLBroadcast(lineBytes: Int, numTrackers: Int = 4, bufferless: Boolean = fa
...
+        TLMessages.PutFullData    -> true.B,
+        TLMessages.PutPartialData -> true.B,
+        TLMessages.ArithmeticData -> true.B,
+        TLMessages.LogicalData    -> true.B,
+        TLMessages.Get            -> false.B,
+        TLMessages.Hint           -> MuxLookup(in.a.bits.param, Wire(Bool()), Array(
+          TLHints.PREFETCH_READ   -> false.B,
+          TLHints.PREFETCH_WRITE  -> true.B)),
+        TLMessages.AcquireBlock   -> acq_needT,
+        TLMessages.AcquirePerm    -> acq_needT))
```
Comment (hcook) [inline src/main/scala/tilelink/Broadcast.scala:236]: These seem like helper functions that could be defined on the `edge`. They're globally true for all TL1.9 caching adapters that have to deal with these opcode/param types, right?

File: src/main/scala/tilelink/Broadcast.scala [hunk 2]
```diff
@@ -159,119 +184,214 @@ class TLBroadcast(lineBytes: Int, numTrackers: Int = 4, bufferless: Boolean = fa
...
-      out.c.valid := Bool(false)
-      out.e.valid := Bool(false)
+      out.b.ready := true.B
+      out.c.valid := false.B
+      out.e.valid := false.B
     }
   }
 }
 
+// addressBits is after removing the block offset (typically lowest 6 bits for 64 bytes)
```
Comment (hcook) [inline src/main/scala/tilelink/Broadcast.scala:274]: isn't this a "tag"?

Comment (terpstra) [reply src/main/scala/tilelink/Broadcast.scala:274]: Well, you would have to strip sector bits and row bits as well before it becomes a tag. I agree addressBits is a bad name. I could not think of a better one, though. If you have a suggestion, I'll change it.

File: src/main/scala/tilelink/Broadcast.scala [hunk 3]
```diff
@@ -159,119 +184,214 @@ class TLBroadcast(lineBytes: Int, numTrackers: Int = 4, bufferless: Boolean = fa
...
+        probe_todo  := todo
+        probe_line  := filter.io.response.bits.address
+        probe_perms := Mux(filter.io.response.bits.needT, TLPermissions.toN, TLPermissions.toB)
+      }
+
+      // Inform the tracker of the number of probe targets
+      val responseCache = filter.io.response.bits.cacheOH | filter.io.response.bits.allocOH
+      val responseCount = PopCount(todo)
+      val responseMSHR = UIntToOH(filter.io.response.bits.mshr, numTrackers).asBools
+      val sack = filter.io.response.fire() && leaveB && others =/= 0.U
```
Comment (hcook) [inline src/main/scala/tilelink/Broadcast.scala:258]: what is `sack`

File: src/main/scala/tilelink/Broadcast.scala (hunk 1 above)
Comment (terpstra) [reply src/main/scala/tilelink/Broadcast.scala:236]: Sure. Fixed.

File: src/main/scala/tilelink/Broadcast.scala (hunk 3 above)
Comment (terpstra) [reply src/main/scala/tilelink/Broadcast.scala:258]: shared acknowledge.

File: src/main/scala/tilelink/Broadcast.scala (hunk 2 above)
Comment (hcook) [reply src/main/scala/tilelink/Broadcast.scala:258]: `blockAddressBits` or `lineAddressBits` maybe?

Comment (terpstra) [reply src/main/scala/tilelink/Broadcast.scala]: Changed to blockAddressBits.

**Pull request merged**